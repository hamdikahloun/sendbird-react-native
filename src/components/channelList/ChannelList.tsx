import React from 'react';
import {FlatList, StyleSheet, View, Text, SafeAreaView} from 'react-native';
import {OpenChannel} from 'sendbird';
import {ChannelPreview} from '@/components';
import {LoadingState} from '../loadingState';

type ChannelListProps = {
  channels: OpenChannel[];
  onChannelPressed?: (channel: OpenChannel) => void;
  loading: boolean;
  error: string;
};

/**
 * ChannelList - This component show a list of channels, allowing you to select the channel you want to open.
 *
 * See:
 *
 * [Group channels](https://sendbird.com/docs/chat/v3/javascript/guides/group-channel-advanced),
 * [Open channels](https://sendbird.com/docs/chat/v3/javascript/guides/open-channel-advanced)
 * @example
 *
 * <ChannelList
 *  loading={loading}
 *  onChannelPressed={onChannelPressed}
 *  channels={channels}
 *  error={error}
 * />
 */
export const ChannelList = ({
  channels,
  onChannelPressed,
  loading,
  error,
}: ChannelListProps) => {
  const renderItem = ({item}: {item: OpenChannel}) => (
    <ChannelPreview onChannelPressed={onChannelPressed} channel={item} />
  );

  const ListEmptyComponent = () => {
    if (loading && !error?.length) {
      return <LoadingState />;
    }
    if (error) {
      return <Text>{error}</Text>;
    }
    if (!loading && !error && !channels.length) {
      return <Text>Empty List</Text>;
    }
    return <View />;
  };

  return (
    <SafeAreaView style={styles.container}>
      <FlatList
        style={styles.flatList}
        data={channels}
        renderItem={renderItem}
        keyExtractor={item => item.url}
        ListEmptyComponent={<ListEmptyComponent />}
        showsVerticalScrollIndicator={false}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  flatList: {
    margin: 16,
  },
});
