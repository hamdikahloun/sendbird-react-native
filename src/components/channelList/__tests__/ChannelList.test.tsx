import React from 'react';
import {render} from '@/../testing';
import {ChannelList} from '../ChannelList';

describe('ChannelList', () => {
  it('should render correctly', async () => {
    const screen = render(
      <ChannelList channels={[]} loading={false} error="" />,
    ).toJSON();
    expect(screen).toMatchSnapshot();
  });
});
