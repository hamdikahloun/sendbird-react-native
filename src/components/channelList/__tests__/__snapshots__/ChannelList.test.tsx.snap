// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelList should render correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={<ListEmptyComponent />}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "margin": 16,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <Text>
        Empty List
      </Text>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
